# What kind of book should we publish to get into the Amazon best sellers list?

We are a new publisher looking to make our first published book a huge success. We focus on our first book because resources are limited and we can't afford to test the market by publishing multiple books and see what would sell the most. Another important detail is we're limited to publishing our e-books; hence, we won't be considering other medium such as physical books or audiobook.

Our aim is somewhat modest and we are only looking to be included in the Amazon best sellers list. If we can be included in that list, we will be more likely to be included in other best sellers list such as the New York Times. 

In relation to the Amazon Book Best Sellers list, we want to answer the following questions:
1. Which genre is the most popular?
2. What is the average price of books?
3. How many reviews does a book receive on average? Does the average number of reviews change significantly depending on the genre?
4. How high is the rating a book receives on average? Does the average rating change significantly depending on the genre?
5. Are there any other factors aside from genre that makes a book popular?

We'll start by answering the above plus some incidental questions found along the way; we will the follow up by formulating one or two book profiles that could potentially become best sellers.

## Loading, Cleaning and Exploring the Data

### Load Data Set

In [1]:
import pandas as pd

amazon_books = pd.read_csv("Amazon_popular_books_dataset.csv")
print(amazon_books.shape)

# Display both head and tail of data frame
amazon_books

(2269, 40)


asin         ISBN10  answered_questions availability  \
0     0007350813     0007350813                   0    In Stock.   
1     0007513763  9780007513765                   0    In Stock.   
2     0008183988     0008183988                   0          NaN   
3     0008305838     0008305838                   0    In Stock.   
4     0008375526     0008375526                   0    In Stock.   
...          ...            ...                 ...          ...   
2264  B07P5BPVGM            NaN                   0          NaN   
2265  B07P5JBCFL            NaN                   0          NaN   
2266  B07NF7DFS2            NaN                   0          NaN   
2267  B07P67N918            NaN                   0          NaN   
2268  B07PFY8K2N            NaN                   0          NaN   

                                                  brand currency  \
0                                          Emily Brontë      USD   
1                                          Drew Daywalt      USD   
2                                      Bernard Cornwell      USD   
3                                        David Walliams      USD   
4                                       Caroline Hirons      USD   
...                                                 ...      ...   
2264                                        Jess Lourey      USD   
2265  Heidi Murkoff (Author, Narrator), Meeghan Hola...      USD   
2266  Clea Shearer (Author, Narrator), Joanna Teplin...      USD   
2267  Lisa Jewell (Author), Tamaryn Payne (Narrator)...      USD   
2268  Elton John (Author, Narrator), Taron Egerton (...      USD   

     date_first_available                                           delivery  \
0                     NaN  ["FREE delivery Tuesday, December 28 if you sp...   
1                     NaN  ["FREE delivery Tuesday, December 28 if you sp...   
2                     NaN  ["FREE delivery January 4 - 10 if you spend $2...   
3                     NaN  ["FREE delivery Tuesday, December 28 if you sp...   
4                     NaN  ["FREE delivery Tuesday, December 28","Or fast...   
...                   ...                                                ...   
2264                  NaN                                                 []   
2265                  NaN                                                 []   
2266                  NaN                                                 []   
2267                  NaN                                                 []   
2268                  NaN                                                 []   

      department description  ...  upc                                   url  \
0            NaN         NaN  ...  NaN  https://www.amazon.com/dp/0007350813   
1            NaN         NaN  ...  NaN  https://www.amazon.com/dp/0007513763   
2            NaN         NaN  ...  NaN  https://www.amazon.com/dp/0008183988   
3            NaN         NaN  ...  NaN  https://www.amazon.com/dp/0008305838   
4            NaN         NaN  ...  NaN  https://www.amazon.com/dp/0008375526   
...          ...         ...  ...  ...                                   ...   
2264         NaN         NaN  ...  NaN  https://www.amazon.com/dp/B07P5BPVGM   
2265         NaN         NaN  ...  NaN  https://www.amazon.com/dp/B07P5JBCFL   
2266         NaN         NaN  ...  NaN  https://www.amazon.com/dp/B07NF7DFS2   
2267         NaN         NaN  ...  NaN  https://www.amazon.com/dp/B07P67N918   
2268         NaN         NaN  ...  NaN  https://www.amazon.com/dp/B07PFY8K2N   

     video  video_count                                         categories  \
0      NaN            0   ["Books","Literature & Fiction","Genre Fiction"]   
1      NaN            0  ["Books","Children's Books","Literature & Fict...   
2      NaN            0   ["Books","Literature & Fiction","Genre Fiction"]   
3      NaN            0  ["Books","Children's Books","Literature & Fict...   
4      NaN            0  ["Books","Crafts, Hobbies & Home","Home I

2269 rows of the data frame matched the data set [documentation](https://github.com/luminati-io/Amazon-popular-books-dataset#amazon-popular-books-dataset):

> This Amazon dataset contains 2269 best-selling books.

On the other hand, the documentation does not mention the number of columns. We should investigate the column names to make sure nothing odd happened when we loaded the data set.

In [2]:
# Show all columns
pd.DataFrame({'columns': amazon_books.columns})

columns
0                   asin
1                 ISBN10
2     answered_questions
3           availability
4                  brand
5               currency
6   date_first_available
7               delivery
8             department
9            description
10              discount
11                domain
12              features
13           final_price
14                format
15             image_url
16          images_count
17         initial_price
18           item_weight
19          manufacturer
20          model_number
21          plus_content
22    product_dimensions
23                rating
24         reviews_count
25          root_bs_rank
26             seller_id
27           seller_name
28             timestamp
29                 title
30                   upc
31                   url
32                 video
33           video_count
34            categories
35     best_sellers_rank
36         buybox_seller
37                 image
38     number_of_sellers
39                colors

### Column Selection

Observing the above results, the column names seems reasonable. Now, we'll need to select the relevant columns because even if keeping irrelevant columns don't have any impact in the results of our analysis, it does clutter the display and also slows down data processing and calculations. 

The documentation mentions that the following are the key columns:

> Key data points included in this free dataset:
> - ASIN
> - ISBN10
> - Categories
> - Reviews count
> - Avg. rating
> - Number of sellers
> - URL
> - Image
> - Final price
> - Title
> - Description
> - Availability

The above columns should be enough for the purpose of this project thus we will restrict our data frame to only those columns for now; we can always add the removed columns later if there's a need for them.

| Full Name         | Column Name       |
|:------------------|:------------------|
| ASIN              | asin              |
| ISBN10            | ISBN10            |
| Categories        | categories        |
| Reviews count     | reviews_count     |
| Average rating    | rating            |
| Number of sellers | number_of_sellers |
| URL               | url               |
| Image             | image             |
| Final price       | final_price       |
| Title             | title             |
| Description       | description       |
| Availability      | availability      |
| Format            | format            |
| Author            | brand             |

We've also added the columns for Format and Author (which is called `brand` - odd, but makes sense because sometimes the book is authored collectively as an institution or publisher; and yes, authors are themselves brands in a way).

In [3]:
# Define the columns to keep
keep_cols = ['asin', 
             'ISBN10', 
             'categories', 
             'reviews_count', 
             'rating', 
             'number_of_sellers', 
             'url', 
             'image', 
             'final_price',
             'title',
             'description',
             'availability',
             'format',
             'brand'
            ]

# Review
amazon_books_updated = amazon_books[keep_cols]
amazon_books_updated.head()

asin         ISBN10  \
0  0007350813     0007350813   
1  0007513763  9780007513765   
2  0008183988     0008183988   
3  0008305838     0008305838   
4  0008375526     0008375526   

                                          categories  reviews_count  \
0   ["Books","Literature & Fiction","Genre Fiction"]          13451   
1  ["Books","Children's Books","Literature & Fict...          16628   
2   ["Books","Literature & Fiction","Genre Fiction"]          11275   
3  ["Books","Children's Books","Literature & Fict...          15520   
4  ["Books","Crafts, Hobbies & Home","Home Improv...          10884   

               rating  number_of_sellers  \
0  4.6 out of 5 stars                NaN   
1  4.8 out of 5 stars                NaN   
2  4.8 out of 5 stars                NaN   
3  4.8 out of 5 stars                NaN   
4  4.8 out of 5 stars                NaN   

                                    url image  final_price  \
0  https://www.amazon.com/dp/0007350813   NaN         3.99   
1  https://www.amazon.com/dp/0007513763   NaN        12.08   
2  https://www.amazon.com/dp/0008183988   NaN          NaN   
3  https://www.amazon.com/dp/0008305838   NaN        20.43   
4  https://www.amazon.com/dp/0008375526   NaN        28.89   

                                               title description availability  \
0               Wuthering Heights (Collins Classics)         NaN    In Stock.   
1                          THE DAYS THE CRAYONS QUIT         NaN    In Stock.   
2        War Lord: Book 13 (The Last Kingdom Series)         NaN          NaN   
3  Code Name Bananas: The hilarious and epic new ...         NaN    In Stock.   
4  Skincare: The award-winning ultimate no-nonsen...         NaN    In Stock.   

                                              format             brand  
0  [{"name":"Kindle","price":"$0.99","url":"/Wuth...      Emily Brontë  
1  [{"name":"Kindle","price":"$10.99","url":"/Day...      Drew Daywalt  
2  [{"name":"Kindle","price":"$11.99","url":"/War...  Bernard Cornwell  
3  [{"name":"Paperback","price":"$24.78","url":"/...    David Walliams  
4  [{"name":"Kindle","price":"$16.99","url":"/Ski...   Caroline Hirons

### Another Round of Column Selection

The previous column selection is mostly done to retrieve plausible columns for the analysis. Upon closer inspection, there are some columns I would like to remove:

| Full Name    | Column Name  | Reason for Removal                                                                                                                                                                                                                                                                                                                                                                                         |
|:-------------|:-------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ASIN         | asin         | The [Amazon Standard Identification Number](https://en.wikipedia.org/wiki/Amazon_Standard_Identification_Number) (ASIN) is not required. It's sufficient to identify a book by its title.                                                                                                                                                                                                                  |
| ISBN         | ISBN10       | Similar to the reason for ASIN removal, the [International Standard Book Number](https://en.wikipedia.org/wiki/ISBN) (ISBN) is not required.                                                                                                                                                                                                                                                               |
| URL          | url          | The url or link to the Amazon page for the book is not necessary. Our analysis doesn't cover anything that looks into the exact url or the page the url links to.                                                                                                                                                                                                                                          |
| Image        | image        | The image of the book is not needed. It could be useful, but it requires different (and potentially more complicated) data analysis methods which are beyond the scope of this project. A worthy candidate for investigation in our next project.                                                                                                                                                          |
| Description  | description  | Similar to book image, this could be useful but can be too complicated to analyse for the purpose of this project.                                                                                                                                                                                                                                                                                         |
| Availability | availability | We are not concerned with a book's stock availability on Amazon; if the book is included in the best seller list, it's good enough whether it's available or not. Other than that, stock availability data is dynamic and the data set is just a snapshot of a book's availability when the data was collected; we'll need a time series data set of availability if we want to use the data for analysis. |

In [4]:
# Define the columns to remove
remove_cols =   [
                 'asin', 
                 'ISBN10', 
                 'url', 
                 'image', 
                 'description',
                 'availability',
                 ]

# Remove the columns
amazon_books_updated = amazon_books_updated.drop(columns=remove_cols)

# Review
amazon_books_updated.head()

categories  reviews_count  \
0   ["Books","Literature & Fiction","Genre Fiction"]          13451   
1  ["Books","Children's Books","Literature & Fict...          16628   
2   ["Books","Literature & Fiction","Genre Fiction"]          11275   
3  ["Books","Children's Books","Literature & Fict...          15520   
4  ["Books","Crafts, Hobbies & Home","Home Improv...          10884   

               rating  number_of_sellers  final_price  \
0  4.6 out of 5 stars                NaN         3.99   
1  4.8 out of 5 stars                NaN        12.08   
2  4.8 out of 5 stars                NaN          NaN   
3  4.8 out of 5 stars                NaN        20.43   
4  4.8 out of 5 stars                NaN        28.89   

                                               title  \
0               Wuthering Heights (Collins Classics)   
1                          THE DAYS THE CRAYONS QUIT   
2        War Lord: Book 13 (The Last Kingdom Series)   
3  Code Name Bananas: The hilarious and epic new ...   
4  Skincare: The award-winning ultimate no-nonsen...   

                                              format             brand  
0  [{"name":"Kindle","price":"$0.99","url":"/Wuth...      Emily Brontë  
1  [{"name":"Kindle","price":"$10.99","url":"/Day...      Drew Daywalt  
2  [{"name":"Kindle","price":"$11.99","url":"/War...  Bernard Cornwell  
3  [{"name":"Paperback","price":"$24.78","url":"/...    David Walliams  
4  [{"name":"Kindle","price":"$16.99","url":"/Ski...   Caroline Hirons

The final list of columns will be the following:

| Full Name         | Column Name       |
|:------------------|:------------------|
| Categories        | categories        |
| Reviews count     | reviews_count     |
| Average rating    | rating            |
| Number of sellers | number_of_sellers |
| Final price       | final_price       |
| Title             | title             |
| Format            | format            |
| Author            | brand             |

### Handling missing values

Let's check for missing values.

In [5]:
# Show the number of missing values per column
amazon_books_updated.isnull().sum()

categories              0
reviews_count           0
rating                  0
number_of_sellers    2265
final_price           877
title                   0
format                 96
brand                   1
dtype: int64

It seems that the number of sellers, final price, format, and brand columns have missing values. The first two, especially  number of sellers, have startling number of missing values. Next, I'll go through each column one by one to evaluate how to handle them.

**Missing values in the number of sellers column**

There's a significant number of missing values in this column. Interesting enough, there are 3 rows with a value.

In [6]:
# Show frequency of values in number_of_sellers
amazon_books_updated['number_of_sellers'].value_counts()

# Show rows has values in number_of_sellers
amazon_books_updated[amazon_books_updated['number_of_sellers'].notnull()]

categories  reviews_count  \
664   ["Books","Humor & Entertainment","Puzzles & Ga...          37782   
1194  ["Books","Self-Help","Relationships","Love & R...          15377   
2029  ["Toys & Games","Learning & Education","Electr...          18477   
2177             ["Books","Calendars","Animals","Dogs"]          11249   

                  rating  number_of_sellers  final_price  \
664   4.9 out of 5 stars               94.0        27.93   
1194  4.7 out of 5 stars               19.0         8.01   
2029  4.8 out of 5 stars                4.0        17.99   
2177  4.8 out of 5 stars                2.0        16.99   

                                                  title format  \
664              Player's Handbook (Dungeons & Dragons)    NaN   
1194  Knock Knock What I Love about You Fill in the ...    NaN   
2029                    VTech Musical Rhymes Book, Pink    NaN   
2177   Pooping Pooches White Elephant Gag Gift Calendar    NaN   

                     brand  
664   Wizards of the Coast  
1194           Knock Knock  
2029                 VTech  
2177       Pooping Pooches

It doesn't seem that we can glean anything important from the available values for number of sellers. Doing the work ourself by filling in the number of sellers is not an option because we do not know exactly when the data was collected; even if we do know, it's unclear whether it's possible to find past data for the known date. Our only option is to drop the number of sellers column though not reluctantly because it is not too important for our analysis.

In [7]:
amazon_books_updated.drop(columns=['number_of_sellers'], inplace=True)
# Verify column has been dropped
pd.DataFrame({"columns": amazon_books_updated.columns})

columns
0     categories
1  reviews_count
2         rating
3    final_price
4          title
5         format
6          brand

**Missing values in the final price and format columns**

We put both the final price and format columns together in this section because they are actually related. Let's see what we mean by that:

In [8]:
# Print the frequency of values in both the final price and format columns
print(amazon_books_updated['final_price'].value_counts(dropna=False))

# Be ready for an eyesore
print(amazon_books_updated['format'].value_counts(dropna=False).head())

NaN      877
9.99      25
6.99      24
8.99      23
15.99     21
        ... 
17.26      1
25.01      1
21.06      1
58.88      1
9.48       1
Name: final_price, Length: 733, dtype: int64
NaN                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                96
[{"name":"Kindle","price":"$0.00","url":"/Picture-Dorian-Gray-AmazonClassics-ebook/dp/B071HDXV91"},{"n

Note that the format column also has prices in them but each price is linked to a certain format. This means that a book might have multiple prices depending on its format.

Other than that, the values in the final price column is as expected at least on a cursory look, but the values in the format column needs further inspection. Before that, we need to do some data conversion for the format column because even though the above looks like a list, it is actually a string. Converting all the values into a list is our next step.

In [9]:
import ast
import numpy as np

'''
**Convert each value that is not null into a list**

`ast.literal_eval` evaluates and converts the literal string into the appropriate Python object. 

The [documentation](https://docs.python.org/3/library/ast.html#ast.literal_eval) mentioned that it is safe-ish; it prevents Python code from being called but depending on the input, it's possible to trigger memory or C-stack exhaustion.

We trust the collector of the data set enough to use `ast.literal_eval` in spite of the potential attack.
'''
amazon_books_updated['format'] = amazon_books_updated['format'].apply(lambda x: ast.literal_eval(x) if not pd.isna(x) else np.nan)

# Making sure that all values in format is either a list or a null value
print("Formats are either list or nulls: {}".format(amazon_books_updated['format'].apply(lambda x: type(x) is list or x is np.nan).all()))

# Verify if the number of nulls is still the same
amazon_books_updated['format'].isnull().sum()

Formats are either list or nulls: True


96

In [10]:
# Is everything a list?
# for val in amazon_books_updated['format']:
# #     print(type(val))
#     if val is list:
#         print(val)

#     if val == False:
#         print(val)

Next, let's explore the possible values for the format name and price.

In [11]:
'''
The value in the format column is a list containing dictionaries. 

In each list is a dictionary that follows the following structure:
{
    "name": "Format",
    "price": "$0.00",
    "url": "/url/to/the/book/page/on/Amazon"
}

We would like to extract the name and price for each book. 

''' 

price_set = set()
format_set = set()

not_null_format = amazon_books_updated[amazon_books_updated['format'].notnull()]['format']

# Loop through the rows with no missing formats and add to the corresponding sets
for formats in not_null_format:
    for book_format in formats:
        format_name = book_format['name']
        price = book_format['price']
        
        format_set.add(format_name)
        price_set.add(price)

        
# Prepare for display
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "last_expr"

format_df = pd.DataFrame({"format": list(format_set)})
display(format_df)

price_df = pd.DataFrame({"price": list(price_set)})
price_df

format
0                            MP3 CD
1                        Flexibound
2                         Paperback
3               Kindle & comiXology
4   Kindle Edition with Audio/Video
5                   Library Binding
6                      Plastic Comb
7                     Multimedia CD
8                        Loose Leaf
9                          Rag Book
10         Audio CD Library Binding
11                           Comics
12                 Textbook Binding
13                        Audiobook
14                    Leather Bound
15                      Sheet music
16                Perfect Paperback
17                  Book Supplement
18         School & Library Binding
19                       Ring-bound
20                       Board book
21                     Spiral-bound
22            Single Issue Magazine
23                          Digital
24                   Product Bundle
25                         Audio CD
26            Mass Market Paperback
27                      Vinyl Bound
28                           Kindle
29                              Map
30                              Toy
31   Preloaded Digital Audio Player
32                            Cards
33                        Hardcover
34              Accessory with book
35                  Audio, Cassette

price
0           
1     $51.95
2     $24.52
3     $15.51
4      $6.65
...      ...
1398  $13.19
1399  $14.97
1400  $23.36
1401   $2.83
1402  $12.18

[1403 rows x 1 columns]

One thing for certain is we didn't expect the variety of book formats (which must be obvious to everyone else but not us). This introduces slight complexity on our part because we need to select the format that best represents our more general book format: physical book and e-book.

On the other hand, the values for price looks reasonable except for the empty value.

Let's identify which book format to keep.

**Identifying relevant formats**

Physical books on Amazon can come in many formats. After reviewing the books we intend to publish, we decided to focus on the following physical formats: paperback and hardcover. For e-books, the obvious choice would be Kindle and we ignore its derivatives which are Kindle & comixology and Kindle Edition with Audio/Video.

We've pruned the format list to only represent what we intend to publish:

| Format Name           | Definition                                                                                                                                                                                                                                                                                                            |
|:----------------------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Paperback             | A book with thick paper or paperboard cover. To contrast with the mass market paperback, a paperback is also called "trade paperback" and is regarded as being higher in quality.                                                                                                                                     |
| Perfect Paperback     | "Perfect" might be a misnomer here depending on your knowledge of book publication. A perfect paperback is not about the quality of the paperback but the kind of binding used. Perfect binding relies solely on adhesive and does not use any stitches.                                                                         |
| Mass Market Paperback | A mass market paperback is similar to a "trade" paperback, but it tends to be smaller and less durable. It's also generally cheaper since it's intended to be produced "in mass". This format is generally used when publishing very popular books in order to make it accessible, mostly price-wise, to most people. |
| Hardcover             | A book with rigid/hard protective covers. A book will typically have a hardcover format which are published early in the publication cycle and only for a limited amount of time. After sales are slower, publisher of the hardcover book normally transition to the cheaper paperback format to reinvigorate sales.  |
| Kindle                | An e-book meant for reading with an Amazon Kindle.                                                                                                                                                                                                                                                                    |

You can find some references on book formats in the <a href="#References">References</a> section.

**Keep only the relevant book formats in the data set**



In [13]:
keep_format = ['Paperback', 'Kindle', 'Perfect Paperback', 'Mass Market Paperback', 'Hardcover']

def keep_specific_format(format_list, format_to_keep):
    if format_list == np.nan:
        return format_list
    for f_dict in format_list:
        if f_dict['name'] == format_to_keep:
            return f_dict['price']

def keep_paperback_hardcover_kindle(row):
#   print(row['format'])
    book_formats = row['format']
    if book_formats is np.nan:
         return row
    for format in book_formats:
        name = format['name']
        price = format['price']
        if name in keep_format:
            split_name = name.split(' ')
            snakecase_name = "_".join(split_name).lower()
            row["{}_price".format(snakecase_name)] = price
    return row

amazon_books_updated = amazon_books_updated.apply(keep_paperback_hardcover_kindle, axis=1)
amazon_books_updated.head()

brand                                         categories  \
0      Emily Brontë   ["Books","Literature & Fiction","Genre Fiction"]   
1      Drew Daywalt  ["Books","Children's Books","Literature & Fict...   
2  Bernard Cornwell   ["Books","Literature & Fiction","Genre Fiction"]   
3    David Walliams  ["Books","Children's Books","Literature & Fict...   
4   Caroline Hirons  ["Books","Crafts, Hobbies & Home","Home Improv...   

   final_price                                             format  \
0         3.99  [{'name': 'Kindle', 'price': '$0.99', 'url': '...   
1        12.08  [{'name': 'Kindle', 'price': '$10.99', 'url': ...   
2          NaN  [{'name': 'Kindle', 'price': '$11.99', 'url': ...   
3        20.43  [{'name': 'Paperback', 'price': '$24.78', 'url...   
4        28.89  [{'name': 'Kindle', 'price': '$16.99', 'url': ...   

  hardcover_price kindle_price mass_market_paperback_price paperback_price  \
0          $10.99        $0.99                       $4.95             NaN   
1           $9.19       $10.99                         NaN             NaN   
2          $18.29       $11.99                         NaN             NaN   
3             NaN          NaN                         NaN          $24.78   
4             NaN       $16.99                         NaN             NaN   

  perfect_paperback_price              rating  reviews_count  \
0                     NaN  4.6 out of 5 stars          13451   
1                     NaN  4.8 out of 5 stars          16628   
2                     NaN  4.8 out of 5 stars          11275   
3                     NaN  4.8 out of 5 stars          15520   
4                     NaN  4.8 out of 5 stars          10884   

                                               title  
0               Wuthering Heights (Collins Classics)  
1                          THE DAYS THE CRAYONS QUIT  
2        War Lord: Book 13 (The Last Kingdom Series)  
3  Code Name Bananas: The hilarious and epic new ...  
4  Skincare: The award-winning ultimate no-nonsen...

It seems that the information from the format column is much richer than the final price column. Another thing we recently noticed is we don't have an idea which format the final price is referring to. This highlights another problem: formats are needed because we're only interested in books that were published at least as a hardcover, paperback, or an e-book; without the format, we risk addding non-representative books into our analysis. 

Because of the above stated reasons, we are considering removing the final price column and also any books with missing formats. Let's investigate those books with no formats first before removing anything.

In [27]:
# Get books with missing formats
no_formats = amazon_books_updated[amazon_books_updated['format'].isnull()]

display(no_formats.head(5))

# Get the percentage of books with missing formats; needed to decide whether to drop them
print("Number of books with missing formats: {}".format(len(no_formats)))
print("Percentage of total data set: {:.2%}".format(len(no_formats)/len(amazon_books_updated)))

brand  \
148                                          Mrs Hinch   
173                                          Dr. Seuss   
186                                       Roger Priddy   
244  by Wynn Kapit (Author), Lawrence M. Elson (Aut...   
273                                  Ann Whitford Paul   

                                            categories  final_price format  \
148  ["Books","Crafts, Hobbies & Home","Home Improv...        19.70    NaN   
173  ["Books","Children's Books","Literature & Fict...        11.98    NaN   
186  ["Books","Children's Books","Education & Refer...        13.17    NaN   
244  ["Books","New, Used & Rental Textbooks","Medic...          NaN    NaN   
273  ["Books","Children's Books","Growing Up & Fact...         4.44    NaN   

    hardcover_price kindle_price mass_market_paperback_price paperback_price  \
148             NaN          NaN                         NaN             NaN   
173             NaN          NaN                         NaN             NaN   
186             NaN          NaN                         NaN             NaN   
244             NaN          NaN                         NaN             NaN   
273             NaN          NaN                         NaN             NaN   

    perfect_paperback_price              rating  reviews_count  \
148                     NaN  4.8 out of 5 stars          15127   
173                     NaN  4.8 out of 5 stars          11204   
186                     NaN  4.8 out of 5 stars          11915   
244                     NaN  4.6 out of 5 stars          10129   
273                     NaN  4.8 out of 5 stars          58744   

                                                 title  
148                    Mrs Hinch: The Activity Journal  
173  The Little Blue Box of Bright and Early Board ...  
186  First 100 Board Book Box Set (3 books): First ...  
244                          The Anatomy Coloring Book  
273                       If Animals Kissed Good Night

Number of books with missing formats: 96
Percentage of total data set: 4.23%


## References



https://en.wikipedia.org/wiki/Hardcover

https://www.julesbuono.com/paperback-vs-mass-market-paperback/

https://en.wikipedia.org/wiki/Paperback#Mass-market

https://en.wikipedia.org/wiki/Amazon_Kindle

https://sellercentral.amazon.com/forums/t/what-is-a-perfect-paperback-in-listing-a-book/236479

https://en.wikibooks.org/wiki/Bookbinding/Perfect_binding